File: 2023/day1.js
const fs = require('node:fs')

const textToNum = {
  'one': "1",
  'two': "2",
  'three': "3",
  'four': "4",
  'five': "5",
  'six': "6",
  'seven': "7",
  'eight': "8",
  'nine': "9",
}

function isNumber(char) {
  return !isNaN(parseInt(char))
}

function reverseText(text) {
  return text.split('').reverse().join('')
}

const reversedKeys = Object.keys(textToNum).map(k => {
  return reverseText(k)
})

const reversedTextToNum = {}

reversedKeys.forEach((k, i) => {
  reversedTextToNum[k] = i + 1
})

// 1. main function that reads the file and processes each line and holds the total sum
function main(file) {
  fs.readFile(file, 'utf-8', (err, data) => {
    if (err) {
      console.error(err);
      return
    }

    // split at new line
    const splitData = data.split('\n')
    const lineValues = []

    splitData.forEach(dat => {
      // const convertedData = convertTextToNum(dat)
      const lineValue = getNumber(dat)
      lineValues.push(lineValue)
    })

    const total = lineValues.reduce((a,b) => a+b)
    console.log("total: ", total);
  })

}

function getNumber(text) {
  let targetNum = ''
  
  let firstNumber;
  let lastNumber;

  for (let i = 0; i < text.length; i++) {
    const char = text[i];
    if (isNumber(char)) {
      firstNumber = char
    } else {
      Object.keys(textToNum).forEach(k => {
        if (text.substring(i).startsWith(k)) {
          firstNumber = textToNum[k]
        }
      })     
    }
    if (firstNumber) {
      break
    }
  }

  const reversedText = text.split('').reverse().join('')

  for (let i = 0; i < reversedText.length; i++) {
    const char = reversedText[i];
    if (isNumber(char)) {
      lastNumber = char
    } else {
      Object.keys(reversedTextToNum).forEach(k => {
        if (reversedText.substring(i).startsWith(k)) {
          lastNumber = reversedTextToNum[k]
        }
      })     
    }
    if (lastNumber) {
      break
    }
  }
  
  targetNum = firstNumber + lastNumber 
  return parseInt(targetNum)
}

main('./input.txt')

File: 2023/day1.rb

def main(f)
  file = File.open(f)
  file_data = file.readlines.map(&:chomp)
  line_values = []

  file_data.each do |d|
    line_val = get_number(d)
    line_values.push(line_val)
  end

  total = line_values.sum
  puts total
end

def string_is_integer?(str)
  Integer(str)
  true
rescue ArgumentError
  false
end

def get_number(text)
  text_to_num = {
    'one' => "1",
    'two' => "2",
    'three' => "3",
    'four' => "4",
    'five' => "5",
    'six' => "6",
    'seven' => "7",
    'eight' => "8",
    'nine' => "9",
  }
  target_number = ''

  for i in 0...text.length
    char = text[i]
    if string_is_integer?(char) 
      first_number = char
    else
      text_to_num.each do |key, value|
        if text[i..-1].start_with?(key)
          first_number = value 
        end
      end
    end
    if first_number
      break
    end
  end

  reversed_text = text.reverse

  reversed_text_to_num = text_to_num.reduce({}) do |new_hash, (k,v)|
    reversed_k = k.reverse
    new_hash[reversed_k] = v
    new_hash
  end

  for i in 0...reversed_text.length
    char = reversed_text[i]
    if string_is_integer?(char) 
      last_number = char
    else
      reversed_text_to_num.each do |key, value|
        if reversed_text[i..-1].start_with?(key)
          last_number = value 
        end
      end
    end
    if last_number 
      break
    end
  end

  target_number = first_number + last_number
  target_number.to_i
end

main("input.txt")